pandas/_libs/lib.pyx

```diff
@@ -2869,8 +2877,8 @@ def map_infer_mask(
 @cython.boundscheck(False)
 @cython.wraparound(False)
 def _map_infer_mask(
-        ndarray[uint8_int64_object_t] out,
-        ndarray[object] arr,
+        ndarray out,
+        ndarray arr,
```

Comment:
This will make iteration slower.

Look into numpy C API iteration functions.

(e.g. look at Validator._validate)

If this gives perf improvement, maybe can be split out.